core_matcher.go

```diff
@@ -52,6 +52,12 @@ func (m *coreMatcher) fields() *coreFields {
 // addPattern - the patternBytes is a JSON text which must be an object. The X is what the matcher returns to indicate
 // that the provided pattern has been matched. In many applications it might be a string which is the pattern's name.
 func (m *coreMatcher) addPattern(x X, patternJSON string) error {
+	return m.addPatternWithPrinter(x, patternJSON, &nullPrinter{})
+}
+
+// addPatternWithPrinter can be called from debugging and under-development code to allow viewing pretty-printed
+// NFAs
+func (m *coreMatcher) addPatternWithPrinter(x X, patternJSON string, printer printer) error {
```

Comment:
since this is taking in an `interface` type (`printer`), IMHO that is the allocation/escaping issue raised in the Youtube link I shared, no?

Reply:
Ahh, I get it. I confess, I didn't watch the YouTube, I just went for obvious allocation.  Now I have, and I learned all sorts of stuff.  In particular `go build -gcflags=-m` and `treating.AllocsPerRun`. Thanks for that!

Now… it’s complicated.  If we turn printer into a struct type where all the methods start 
```go
if !pp.enabled() {
  return constant;
}
// else do real work
```
then…
1. there is one less allocation for any method that has a `printer` argument
2. each ``printer.anything()`` call actually executes the conditional above
3. but it's a direct call.

On the other hand, if we use the current interface approach:
1. There is an allocation wherever the method has a printer argument.
2. but in the nullPrinter case there's no conditional executed, it just does `return`.
3. but it's a virtual call so runtime has to do a lookup

Other factors…
1. This is only called on `addPattern` calls which are much rarer than `patternsForEvent` calls
2. The `addPattern` call is fairly expensive, does lots of computation and allocation
3. The allocation is about the minimum possible, an empty structure and dispatch information

The two most significant factors in my mind are both intuitions unsupported by data:
1. The effect is small and will not  be significant in the context of the heavyweight `addPattern` logic
2. Of the effects, the struct call vs the indirect interface call might actually be a big deal.

What do you think?

Reply:
Thx for clarifying that this is only on the addPattern code path - missed that. Then I completely agree that interface approach works bc we are not in the data path 👍